### ISSUE TITLE
FR: Add `--enforce_repository` flag

### ISSUE THREAD

**Issue by damienmg** [bazelbuild/bazel#1820]
This flag would add a repository to the workspace file from the command line. If the new repository has the same name than another one in the WORKSPACE file, the one in the WORKSPACE file will be ignored.

This is useful for continuous integration use cases when we want to test change to downstream dependency (so we can inject workspace dependency). It is also useful for IDE integration to insert skylark aspect in the package path. The current approach of using `--package_path` regenerate spurious invalidation.

An example syntax would be:

```
bazel build --enforce_repository="new_local_repository(name = 'foo', path = 'bar')" //...
```

**Adding/Removing that flag should not invalidate target that does not depends on it**

**Comment by kchodorow** [comment]
It seems like the feature request is probably "have a way of overriding external repositories for CIs, users in China, people on an intranet, etc."  A flag that takes a python string doesn't seem like a great way to do that, there should at least be a design doc.

**Comment by lberki** [comment]
+1 for Kristina's opinion; I don't think the command line option is a good idea, because you are limited to short strings and are plagued by quoting issues.

How about adding a `--workspace_prefix` or `--workspace_suffix` option that would be prepended/appended to the workspace file? It would also be useful for cases like individual developers wanting to override parts of the `WORKSPACE` file they check out.

**Comment by damienmg** [comment]
The original idea was `--enforce_repositories` that takes a workspace file (we discussed offline back then that since we want that to prevent other workspace definition from loading, it prefix/suffix is not the good name). But we had some discussion back in July and @ulfjack suggested we have a synthetic way to express it from the command line, I don't remember why (you were probably not in that meeting). I don't want to introduce a second language for describing repository so we ended up putting the python definition in the command line.

I don't care about it being `--enforce_repositories` or `--enforce_repository` but we need to decide before starting to work on it :)

**Comment by lberki** [comment]
+1 for not introducing Yet Another Language. But putting complex Pythonic expressions in a shell command doesn't sound like a recipe for happiness, either.

**Comment by ulfjack** [comment]
Automatically rewriting the WORKSPACE file, which can contain arbitrarily
complex Skylark code, doesn't sound like a recipe for success. For CI, I
think we mostly need a way to override remote locations specifically, right?

**Comment by damienmg** [comment]
@katre: per discussion today, you get to write a doc about that feature and decide on the good solution :)

**Comment by damienmg** [comment]
Yes that's what this proposal is for. It is specifying a workspace file or
part of it from the command line that would take precedence over the same
repositories defined in the WORKSPACE file itself. It won't implies any
rewriting of the WORKSPACE file.

The CI system would assume that the name uniquely identify a dependencies
that is what we have been pushing forward. I guess @katre has all the
inputs now :) Let's continue the discussion on a specific design doc.

**Comment by googlestanke** [comment]
Suggestion: the flag's purpose would be more self-evident if we name it "override_dependency"

**Comment by kchodorow** [comment]
Fixed by dfcd5da86e2acfd42ca09c7f65e012465ab3e382.
